ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Forge.hs

```diff
@@ -61,7 +73,50 @@ forgeShelleyBlock hotKey canBeLeader cfg curNo curSlot tickedLedger txs isLeader
     TPraosConfig { tpraosParams = TPraosParams { tpraosSlotsPerKESPeriod } } =
       configConsensus cfg
 
-    body = SL.toTxSeq @era $ Seq.fromList $ (\(ShelleyValidatedTx _ tx) -> tx) <$> txs
+    computedMaxTxCapacity = computeMaxTxCapacity tickedLedger maxTxCapacityOverride
+    computedMaxTxExUnitCapacity = maxTxExUnitCapacity tickedLedger
+
+    body =
+        SL.toTxSeq @era
+      . Seq.fromList
+      . toTxInBlock
+      $ toPayloads txs
+
+    toPayloads :: [Validated (GenTx (ShelleyBlock era))] -> BlockPayloads era
+    toPayloads = foldr (appendTx computedMaxTxCapacity) initBlockPayloads
```

Comment:
Same issue here, with `foldr` as in the Byron ledger: this is taking suffixes.

My advice for both is to right a `go` list, not use `foldr` or `foldl` etc.